File: scripts/diff_summary.py
#!/usr/bin/env python3
"""Layer 1 (language-agnostic) code-review distiller.

Parses a `git diff` into a compact, structured JSON summary so a reviewing
agent can prioritise *where* to look without reading the full patch into its
context window.

Output (stdout) is a single JSON object:

    {
      "source": "range" | "staged" | "file",
      "range": "main..HEAD",                # when source == range
      "totals": {"files", "additions", "deletions", "hunks"},
      "files": [ { ...per-file... } ],
      "risk_flags": [ "code_changed_without_tests", ... ],   # repo-wide
      "review_order": [ "path", ... ]       # highest risk first
    }

Per-file object:

    {
      "path", "old_path"?, "status",        # added|modified|deleted|renamed
      "language", "additions", "deletions",
      "is_test", "is_generated", "is_binary",
      "hunks": [ {"old_start","old_lines","new_start","new_lines",
                  "header","added","deleted"} ],
      "risk_flags": [ "large_hunk", "file_deleted", ... ]
    }

Dependencies: Python 3 + git only. Works for any language.
"""

import argparse
import json
import os
import re
import subprocess
import sys

# --- Heuristics ------------------------------------------------------------

# A hunk touching this many lines (added + deleted) is flagged large_hunk.
LARGE_HUNK_LINES = 80
# A file changing this many lines total is flagged large_file_change.
LARGE_FILE_LINES = 300

TEST_PATTERNS = [
    re.compile(r"(^|/)tests?/"),
    re.compile(r"(^|/)__tests__/"),
    re.compile(r"(^|/)spec/"),
    re.compile(r"\.(test|spec)\.[^/]+$"),
    re.compile(r"_test\.[^/]+$"),
    re.compile(r"(^|/)test_[^/]+$"),
    re.compile(r"\.feature$"),
]

GENERATED_PATTERNS = [
    re.compile(r"\.min\.(js|css)$"),
    re.compile(r"(^|/)(dist|build|out|vendor|node_modules)/"),
    re.compile(r"package-lock\.json$"),
    re.compile(r"yarn\.lock$"),
    re.compile(r"pnpm-lock\.yaml$"),
    re.compile(r"Cargo\.lock$"),
    re.compile(r"poetry\.lock$"),
    re.compile(r"go\.sum$"),
    re.compile(r"\.pb\.go$"),
    re.compile(r"_pb2\.py$"),
    re.compile(r"\.snap$"),
    re.compile(r"(^|/)generated/"),
    re.compile(r"\.generated\.[^/]+$"),
]

# Extension -> language label. Kept small and obvious; "unknown" otherwise.
EXT_LANG = {
    "ts": "typescript", "tsx": "typescript",
    "js": "javascript", "jsx": "javascript", "mjs": "javascript", "cjs": "javascript",
    "py": "python",
    "go": "go",
    "rs": "rust",
    "java": "java",
    "kt": "kotlin",
    "rb": "ruby",
    "php": "php",
    "c": "c", "h": "c",
    "cc": "cpp", "cpp": "cpp", "cxx": "cpp", "hpp": "cpp",
    "cs": "csharp",
    "swift": "swift",
    "scala": "scala",
    "sh": "shell", "bash": "shell",
    "json": "json", "yaml": "yaml", "yml": "yaml",
    "md": "markdown",
    "sql": "sql",
}

CODE_LANGS = {
    "typescript", "javascript", "python", "go", "rust", "java", "kotlin",
    "ruby", "php", "c", "cpp", "csharp", "swift", "scala", "shell",
}


def matches_any(path, patterns):
    return any(p.search(path) for p in patterns)


def guess_language(path):
    _, ext = os.path.splitext(path)
    return EXT_LANG.get(ext.lstrip(".").lower(), "unknown")


# --- Diff parsing ----------------------------------------------------------

HUNK_RE = re.compile(
    r"^@@ -(?P<os>\d+)(?:,(?P<ol>\d+))? \+(?P<ns>\d+)(?:,(?P<nl>\d+))? @@(?P<ctx>.*)$"
)


def parse_diff(text):
    """Parse unified `git diff` text into a list of per-file dicts."""
    files = []
    current = None
    in_hunk = False

    lines = text.splitlines()
    i = 0
    while i < len(lines):
        line = lines[i]

        if line.startswith("diff --git "):
            if current is not None:
                files.append(current)
            current = _new_file_record()
            in_hunk = False
            # paths from the "diff --git a/x b/y" header (fallback)
            m = re.match(r"diff --git a/(.+?) b/(.+)$", line)
            if m:
                current["_a"] = m.group(1)
                current["_b"] = m.group(2)
            i += 1
            continue

        if current is None:
            i += 1
            continue

        if line.startswith("old mode") or line.startswith("new mode"):
            i += 1
            continue
        if line.startswith("new file mode"):
            current["status"] = "added"
            i += 1
            continue
        if line.startswith("deleted file mode"):
            current["status"] = "deleted"
            i += 1
            continue
        if line.startswith("rename from "):
            current["old_path"] = line[len("rename from "):]
            current["status"] = "renamed"
            i += 1
            continue
        if line.startswith("rename to "):
            current["path"] = line[len("rename to "):]
            current["status"] = "renamed"
            i += 1
            continue
        if line.startswith("copy from ") or line.startswith("copy to "):
            i += 1
            continue
        if line.startswith("similarity index") or line.startswith("dissimilarity index"):
            i += 1
            continue
        if line.startswith("index "):
            i += 1
            continue
        if line.startswith("Binary files") or line.startswith("GIT binary patch"):
            current["is_binary"] = True
            i += 1
            continue
        if line.startswith("--- "):
            path = line[4:]
            if path != "/dev/null":
                current["old_path"] = _strip_prefix(path)
            i += 1
            continue
        if line.startswith("+++ "):
            path = line[4:]
            if path != "/dev/null":
                current["path"] = _strip_prefix(path)
            i += 1
            continue

        m = HUNK_RE.match(line)
        if m:
            hunk = {
                "old_start": int(m.group("os")),
                "old_lines": int(m.group("ol")) if m.group("ol") else 1,
                "new_start": int(m.group("ns")),
                "new_lines": int(m.group("nl")) if m.group("nl") else 1,
                "header": m.group("ctx").strip(),
                "added": 0,
                "deleted": 0,
            }
            current["hunks"].append(hunk)
            in_hunk = True
            i += 1
            continue

        if in_hunk and current["hunks"]:
            hunk = current["hunks"][-1]
            if line.startswith("+"):
                hunk["added"] += 1
                current["additions"] += 1
            elif line.startswith("-"):
                hunk["deleted"] += 1
                current["deletions"] += 1
            # context / "\ No newline" lines are ignored
        i += 1

    if current is not None:
        files.append(current)

    return [_finalize_file(f) for f in files]


def _new_file_record():
    return {
        "path": None,
        "old_path": None,
        "status": "modified",
        "additions": 0,
        "deletions": 0,
        "is_binary": False,
        "hunks": [],
        "_a": None,
        "_b": None,
    }


def _strip_prefix(path):
    # git uses a/ and b/ prefixes by default
    if path.startswith("a/") or path.startswith("b/"):
        return path[2:]
    return path


def _finalize_file(f):
    path = f["path"] or f["_b"] or f["old_path"] or f["_a"]
    old_path = f["old_path"]
    f.pop("_a", None)
    f.pop("_b", None)
    f["path"] = path

    # Don't report a redundant old_path unless it actually differs.
    if f["status"] != "renamed" and old_path == path:
        f["old_path"] = None

    f["language"] = guess_language(path or "")
    f["is_test"] = matches_any(path or "", TEST_PATTERNS)
    f["is_generated"] = matches_any(path or "", GENERATED_PATTERNS)

    f["risk_flags"] = compute_file_risk(f)
    return f


# --- Risk scoring ----------------------------------------------------------

def compute_file_risk(f):
    flags = []
    if f["status"] == "deleted":
        flags.append("file_deleted")
    if f["status"] == "renamed":
        flags.append("file_renamed")
    if f["is_binary"]:
        flags.append("binary_change")

    total = f["additions"] + f["deletions"]
    if total >= LARGE_FILE_LINES:
        flags.append("large_file_change")
    if any((h["added"] + h["deleted"]) >= LARGE_HUNK_LINES for h in f["hunks"]):
        flags.append("large_hunk")

    if f["is_generated"] and not f["is_test"]:
        flags.append("generated_file")
    return flags


def file_risk_score(f):
    """Heuristic numeric score for ordering. Higher = look first."""
    score = 0
    weights = {
        "file_deleted": 30,
        "large_file_change": 25,
        "large_hunk": 15,
        "file_renamed": 10,
        "binary_change": 5,
    }
    for flag in f["risk_flags"]:
        score += weights.get(flag, 0)
    score += f["additions"] + f["deletions"]
    # Generated/lock files are noise -> push down.
    if f["is_generated"]:
        score -= 100
    # Test files are usually lower priority to scrutinise for bugs.
    if f["is_test"]:
        score -= 20
    return score


def compute_repo_risk(files):
    flags = []
    code_files = [
        f for f in files
        if f["language"] in CODE_LANGS and not f["is_generated"] and not f["is_test"]
    ]
    test_files = [f for f in files if f["is_test"]]
    if code_files and not test_files:
        flags.append("code_changed_without_tests")
    if any(f["status"] == "deleted" for f in files):
        flags.append("contains_deletions")
    if len(files) >= 20:
        flags.append("large_changeset")
    return flags


# --- Diff acquisition ------------------------------------------------------

def run_git_diff(args):
    cmd = ["git", "diff", "--no-color", "-M"]
    if args.staged:
        cmd.append("--cached")
    if args.range:
        cmd.append(args.range)
    try:
        out = subprocess.run(
            cmd, cwd=args.cwd, check=True,
            stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True,
        )
    except FileNotFoundError:
        sys.stderr.write("error: git not found on PATH\n")
        sys.exit(2)
    except subprocess.CalledProcessError as e:
        sys.stderr.write("error: git diff failed: %s\n" % e.stderr.strip())
        sys.exit(2)
    return out.stdout


def main():
    p = argparse.ArgumentParser(description="Layer 1 language-agnostic diff distiller")
    p.add_argument("--range", help="git revision range, e.g. main..HEAD")
    p.add_argument("--staged", action="store_true", help="diff the staged index")
    p.add_argument("--file", help="read a unified diff from this file instead of git")
    p.add_argument("--cwd", default=".", help="repository working directory")
    args = p.parse_args()

    if args.file:
        with open(args.file, "r", encoding="utf-8", errors="replace") as fh:
            text = fh.read()
        source = "file"
    else:
        text = run_git_diff(args)
        source = "staged" if args.staged else "range"

    files = parse_diff(text)

    ordered = sorted(files, key=file_risk_score, reverse=True)

    result = {
        "source": source,
        "totals": {
            "files": len(files),
            "additions": sum(f["additions"] for f in files),
            "deletions": sum(f["deletions"] for f in files),
            "hunks": sum(len(f["hunks"]) for f in files),
        },
        "files": files,
        "risk_flags": compute_repo_risk(files),
        "review_order": [f["path"] for f in ordered],
    }
    if source == "range" and args.range:
        result["range"] = args.range

    json.dump(result, sys.stdout, indent=2)
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: js/diff_summary.mjs
#!/usr/bin/env node
// Layer 1 (language-agnostic) code-review distiller — Node port.
//
// Pure Node: no dependencies, no build step. Runs anywhere Node is present
// (which is anywhere Claude Code runs). Mirrors scripts/diff_summary.py output.
//
// Usage:
//   node js/diff_summary.mjs --range main..HEAD --cwd <repo>
//   node js/diff_summary.mjs --staged --cwd <repo>
//   node js/diff_summary.mjs --file some.diff
//   add --compact for minified JSON with empty fields dropped.

import { execFileSync } from "node:child_process";
import { readFileSync } from "node:fs";
import path from "node:path";

const LARGE_HUNK_LINES = 80;
const LARGE_FILE_LINES = 300;

const TEST_PATTERNS = [
  /(^|\/)tests?\//,
  /(^|\/)__tests__\//,
  /(^|\/)spec\//,
  /\.(test|spec)\.[^/]+$/,
  /_test\.[^/]+$/,
  /(^|\/)test_[^/]+$/,
  /\.feature$/,
];

const GENERATED_PATTERNS = [
  /\.min\.(js|css)$/,
  /(^|\/)(dist|build|out|vendor|node_modules)\//,
  /package-lock\.json$/,
  /yarn\.lock$/,
  /pnpm-lock\.yaml$/,
  /Cargo\.lock$/,
  /poetry\.lock$/,
  /go\.sum$/,
  /\.pb\.go$/,
  /_pb2\.py$/,
  /\.snap$/,
  /(^|\/)generated\//,
  /\.generated\.[^/]+$/,
];

const EXT_LANG = {
  ts: "typescript", tsx: "typescript",
  js: "javascript", jsx: "javascript", mjs: "javascript", cjs: "javascript",
  py: "python", go: "go", rs: "rust", java: "java", kt: "kotlin",
  rb: "ruby", php: "php", c: "c", h: "c",
  cc: "cpp", cpp: "cpp", cxx: "cpp", hpp: "cpp",
  cs: "csharp", swift: "swift", scala: "scala",
  sh: "shell", bash: "shell",
  json: "json", yaml: "yaml", yml: "yaml", md: "markdown", sql: "sql",
};

const CODE_LANGS = new Set([
  "typescript", "javascript", "python", "go", "rust", "java", "kotlin",
  "ruby", "php", "c", "cpp", "csharp", "swift", "scala", "shell",
]);

const matchesAny = (p, pats) => pats.some((re) => re.test(p));

function guessLanguage(p) {
  const ext = path.extname(p || "").replace(/^\./, "").toLowerCase();
  return EXT_LANG[ext] || "unknown";
}

const HUNK_RE =
  /^@@ -(?<os>\d+)(?:,(?<ol>\d+))? \+(?<ns>\d+)(?:,(?<nl>\d+))? @@(?<ctx>.*)$/;

function stripPrefix(p) {
  return p.startsWith("a/") || p.startsWith("b/") ? p.slice(2) : p;
}

function newFile() {
  return {
    path: null, old_path: null, status: "modified",
    additions: 0, deletions: 0, is_binary: false, hunks: [], _a: null, _b: null,
  };
}

function parseDiff(text) {
  const files = [];
  let cur = null;
  let inHunk = false;
  for (const line of text.split("\n")) {
    if (line.startsWith("diff --git ")) {
      if (cur) files.push(cur);
      cur = newFile();
      inHunk = false;
      const m = line.match(/^diff --git a\/(.+?) b\/(.+)$/);
      if (m) { cur._a = m[1]; cur._b = m[2]; }
      continue;
    }
    if (!cur) continue;
    if (line.startsWith("old mode") || line.startsWith("new mode")) continue;
    if (line.startsWith("new file mode")) { cur.status = "added"; continue; }
    if (line.startsWith("deleted file mode")) { cur.status = "deleted"; continue; }
    if (line.startsWith("rename from ")) { cur.old_path = line.slice(12); cur.status = "renamed"; continue; }
    if (line.startsWith("rename to ")) { cur.path = line.slice(10); cur.status = "renamed"; continue; }
    if (line.startsWith("copy from ") || line.startsWith("copy to ")) continue;
    if (line.startsWith("similarity index") || line.startsWith("dissimilarity index")) continue;
    if (line.startsWith("index ")) continue;
    if (line.startsWith("Binary files") || line.startsWith("GIT binary patch")) { cur.is_binary = true; continue; }
    if (line.startsWith("--- ")) {
      const p = line.slice(4);
      if (p !== "/dev/null") cur.old_path = stripPrefix(p);
      continue;
    }
    if (line.startsWith("+++ ")) {
      const p = line.slice(4);
      if (p !== "/dev/null") cur.path = stripPrefix(p);
      continue;
    }
    const m = line.match(HUNK_RE);
    if (m) {
      cur.hunks.push({
        old_start: +m.groups.os, old_lines: m.groups.ol ? +m.groups.ol : 1,
        new_start: +m.groups.ns, new_lines: m.groups.nl ? +m.groups.nl : 1,
        header: m.groups.ctx.trim(), added: 0, deleted: 0,
      });
      inHunk = true;
      continue;
    }
    if (inHunk && cur.hunks.length) {
      const h = cur.hunks[cur.hunks.length - 1];
      if (line.startsWith("+")) { h.added++; cur.additions++; }
      else if (line.startsWith("-")) { h.deleted++; cur.deletions++; }
    }
  }
  if (cur) files.push(cur);
  return files.map(finalizeFile);
}

function finalizeFile(f) {
  const p = f.path || f._b || f.old_path || f._a;
  const oldPath = f.old_path;
  delete f._a; delete f._b;
  f.path = p;
  if (f.status !== "renamed" && oldPath === p) f.old_path = null;
  f.language = guessLanguage(p || "");
  f.is_test = matchesAny(p || "", TEST_PATTERNS);
  f.is_generated = matchesAny(p || "", GENERATED_PATTERNS);
  f.risk_flags = computeFileRisk(f);
  return f;
}

function computeFileRisk(f) {
  const flags = [];
  if (f.status === "deleted") flags.push("file_deleted");
  if (f.status === "renamed") flags.push("file_renamed");
  if (f.is_binary) flags.push("binary_change");
  const total = f.additions + f.deletions;
  if (total >= LARGE_FILE_LINES) flags.push("large_file_change");
  if (f.hunks.some((h) => h.added + h.deleted >= LARGE_HUNK_LINES)) flags.push("large_hunk");
  if (f.is_generated && !f.is_test) flags.push("generated_file");
  return flags;
}

function fileRiskScore(f) {
  const weights = {
    file_deleted: 30, large_file_change: 25, large_hunk: 15,
    file_renamed: 10, binary_change: 5,
  };
  let score = 0;
  for (const flag of f.risk_flags) score += weights[flag] || 0;
  score += f.additions + f.deletions;
  if (f.is_generated) score -= 100;
  if (f.is_test) score -= 20;
  return score;
}

function computeRepoRisk(files) {
  const flags = [];
  const codeFiles = files.filter((f) => CODE_LANGS.has(f.language) && !f.is_generated && !f.is_test);
  const testFiles = files.filter((f) => f.is_test);
  if (codeFiles.length && !testFiles.length) flags.push("code_changed_without_tests");
  if (files.some((f) => f.status === "deleted")) flags.push("contains_deletions");
  if (files.length >= 20) flags.push("large_changeset");
  return flags;
}

function stripEmpty(obj) {
  if (Array.isArray(obj)) return obj.map(stripEmpty);
  if (obj && typeof obj === "object") {
    const out = {};
    for (const [k, v] of Object.entries(obj)) {
      const sv = stripEmpty(v);
      if (sv === null || (Array.isArray(sv) && sv.length === 0) || sv === "") continue;
      out[k] = sv;
    }
    return out;
  }
  return obj;
}

function emit(result, compact) {
  if (compact) process.stdout.write(JSON.stringify(stripEmpty(result)) + "\n");
  else process.stdout.write(JSON.stringify(result, null, 2) + "\n");
}

function parseArgs(argv) {
  const a = { cwd: ".", range: null, staged: false, file: null, compact: false };
  for (let i = 0; i < argv.length; i++) {
    const x = argv[i];
    if (x === "--range") a.range = argv[++i];
    else if (x === "--staged") a.staged = true;
    else if (x === "--file") a.file = argv[++i];
    else if (x === "--cwd") a.cwd = argv[++i];
    else if (x === "--compact") a.compact = true;
  }
  return a;
}

function runGitDiff(a) {
  const args = ["diff", "--no-color", "-M"];
  if (a.staged) args.push("--cached");
  if (a.range) args.push(a.range);
  try {
    return execFileSync("git", args, { cwd: a.cwd, encoding: "utf8", maxBuffer: 1 << 28 });
  } catch (e) {
    process.stderr.write("error: git diff failed: " + (e.stderr || e.message) + "\n");
    process.exit(2);
  }
}

function main() {
  const a = parseArgs(process.argv.slice(2));
  let text, source;
  if (a.file) { text = readFileSync(a.file, "utf8"); source = "file"; }
  else { text = runGitDiff(a); source = a.staged ? "staged" : "range"; }

  const files = parseDiff(text);
  for (const f of files) f.risk_score = fileRiskScore(f);
  const ordered = [...files].sort((x, y) => y.risk_score - x.risk_score);

  const result = {
    source,
    totals: {
      files: files.length,
      additions: files.reduce((s, f) => s + f.additions, 0),
      deletions: files.reduce((s, f) => s + f.deletions, 0),
      hunks: files.reduce((s, f) => s + f.hunks.length, 0),
    },
    files,
    risk_flags: computeRepoRisk(files),
    review_order: ordered.map((f) => f.path),
  };
  if (source === "range" && a.range) result.range = a.range;
  emit(result, a.compact);
}

main();
